_posts/2024-01-21-comparing-cas-and-karpenter.markdown

```diff
@@ -0,0 +1,273 @@
+to these problems, but it's a point of concern that think about frequently.
+
+In addition to code, the joining of the communities has lead to an advancement
+of defining common APIs around cluster lifecycle in relation to node
+provisioning and removal. See the
+[Cluster Autoscaler/Karpenter API Alignment AEP draft][sig-align] for more
+information.
+
+## Is one "better" than the other?
+
+This is a question that I don't think is quite right. Both CAS and Karpenter
+are tremendous software applications with many satisfied users. They are
+different in approach and features, and I don't think it's fair to proclaim
+that one is greater or lesser than the other. I think it's more appropriate to
+ask, _which one is better for you?_
+
+There is tremendous overlap between the two applications but they do have
+different, opinionated, approaches to solving the problem of having
+just-in-time resources in Kubernetes. In many respects, Karpenter can be
+configured in a manner to perform the same task as CAS, and that cannot be
+said in the converse. In this respect Karpenter's features might be seen as
+a superset CAS's features.
+
+On the other hand, in clusters where swings in cluster size might be
+lower and choice of instance type is not required to be as dynamic, operating
+the CAS could prove to be a less complex task. Especially if you have been
+using the CAS for years already.
+
+Additionally, CAS currently runs on 28 providers listed in the repository, with
+at least 2 of those (Cluster API and OpenStack) being platforms that run on a
+multitude of other platforms. Karpenter currently only supports AWS and Azure.
+
+## Caution, strong opinions ahead
+
+Karpenter seems ideal for situations where you want to manage larger
+heterogeneous clusters that have high amounts of workload churn. Where resource
+maximization and cost reduction are primary drivers to configuring and
+optimizing the cluster. In these scenarios, Karpenter's ability to consolidate
+and choose from a wide variety of instance types will be very beneficial.
+
+The CAS seems well suited in cases where your cluster has a very reliable rate
+of growth, and the instance types are more homogeneous across the cluster. This
+applies well in smaller clusters, especially if the need for scaling is only a
+single node or two over a short period of time (e.g. bursting for an evening).
+It also applies well in clusters where the instances are treated as a large pool
+and the specific types are less important, or in situations where having some
+overprovisioned capacity is preferred.
+
+I find Karpenter's configuration options to be more complicated than the CAS,
+and I'm not sure how bad it is to debug issues with Karpenter as I have much
+more time on the CAS. I really like that Karpenter has taken the approach to
+use API resources (`NodePool`, `NodeClaim`, etc.) to drive its behavior as I
+think it makes it easier to reckon about functionality.
```

Comment:
And it exposes more context to the user, makes it more visible

Reply:
good suggestion